### PR TITLE
First Response Latency Implementation

### PR DESCRIPTION
**Pull request by elopezanaya**
### Description
_Include a description of the problem to be solved_

## Solution Proposed
_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_

## First Response Latency

__Rules of engagement__ : 
- Should only track bot engaged messages
- Start to track only when C2 send a message
- Stop clock once a message is received

__Rules of disengagement__ : 
- Once it detects an Agent Message, should stop tracking , and no record any telemetry
- Do not track system messages ( extended version will support typing message from copilot but not ready yet)
- Do not track history messages
- Reload should not trigger clock engagement unless C2 types

__Rules for starting tracking__:
- Track starts once C2 type a message
- if user send various messages, it will only track the first message
- until there is a response from Copilot, it  will ignore all C2 messages for tracking

__Rules for stopping tracking and save telemetry__ 
- Clock stops once the first message arrives
- If bot sends more than one message, subsequent messages wont be tracked
- Once the cycle is completed , the telemetry will be saved
image

## History message re-definition

During testing it was discovered a improvement to define a history message after Chat Adapter wont longer offer a way to identify this type of messages

image
## Message strategy

New message handler was using a long block to define the logic to handle all messages, by introducing a strategy pattern, now each scenario can handle its own logic , breaking the high couple dependency

## Test cases and evidence
_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

image

image

### Reconnection
image
image
image

### Sanity Tests
-   [ ] You have tested all changes in Popout mode
-   [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
-   [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y
- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

__*Please provide justification if any of the validations has been skipped.*__

### CODE REVIEW

File: chat-widget/src/firstresponselatency/Constants.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+export enum ScenarioType {
+    UserSendMessageStrategy = "UserSendMessageStrategy",
+    SystemMessageStrategy = "SystemMessageStrategy",
+    ReceivedMessageStrategy = "ReceivedMessageStrategy"
+}
+
+export type MessagePayload = {
```
Comment (xTEddie) [inline chat-widget/src/firstresponselatency/Constants.ts:7]: `MessagePayload` is a generic name. It may be reused in other features in the future. Should a prefix be added to be more distinguishable? `FRLMessagePayload`?

File: chat-widget/src/firstresponselatency/Constants.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    channelData?: any;
+    chatId?: string;
+    conversationId?: string;
+    isChatComplete: boolean;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    attachment?: any[];
+};
+
+export type TrackingMessage = {
```
Comment (xTEddie) [inline chat-widget/src/firstresponselatency/Constants.ts:25]: `FRLTrackingMessage`?

File: chat-widget/src/firstresponselatency/FirstResponseLatencyTracker.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    // Tracking Functions
+    private startTracking(payload: MessagePayload): void {
+        //  this prevents to initiate tracking for multiple incoming messages
+        if (this.isStarted) {
+            return;
+        }
+        // this is to ensure we track only messages where bot is engaged
+        if (!this.isABotConversation) {
```
Comment (xTEddie) [inline chat-widget/src/firstresponselatency/FirstResponseLatencyTracker.ts:41]: Shall both `if` conditions be merged?

```
if (this.isStarted || !this.isABotConversation) return;
```

File: chat-widget/src/firstresponselatency/FirstResponseLatencyTracker.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        // The idea  of using types is to enrich telemetry data 
+        this.startTrackingMessage = this.createTrackingMessage(payload, "userMessage");
+    }
+
+    private handleAgentMessage(payload: MessagePayload): void {
+        // this tag so far is only present in agent messages
+        if (payload?.tags?.includes("public")) {
+            this.deregister();
+        }
+        return;
```
Comment (xTEddie) [inline chat-widget/src/firstresponselatency/FirstResponseLatencyTracker.ts:58]: `return` is not needed

File: chat-widget/src/firstresponselatency/util.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                if (isNaN(activityId)) {
+                    return false;
+                }
+
+                // if the activity id is less than the start time, it means that the message is a history message
+                if (activityId < startTime) {
+                    return true;
+                }
+                // anything else will be considered a new message
+                return false;
```
Comment (xTEddie) [inline chat-widget/src/firstresponselatency/util.ts:27]: `return false` is not needed given such statement exists below already

File: chat-widget/src/firstresponselatency/util.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            }
+            return false;
+        }
+    } catch (e) {
+        // if there is an error in parsing the activity id, we will consider it a new message
+        console.error("Error in parsing activity id: ", e);
+        return false;
```
Comment (xTEddie) [inline chat-widget/src/firstresponselatency/util.ts:34]: `return false` is not needed given such statement exists outside the catch block already

File: chat-widget/src/plugins/newMessageEventHandler.ts [hunk 7]
```diff
@@ -1,112 +1,140 @@
...
+    // epoch time in utc for when start to listen.
+    // We dont longer have a mechanism to know if a message is history or new, so any message older than the time we start listening will be considered a history message.
+    // this is a workaround for the fact that we dont have a way to identify if a message is history or new, and it will provide consistency across different scenarios
+    const startTime = (new Date().getTime());
+    let isHistoryMessageReceivedEventRasied = false;
+
     const onNewAdapterActivityHandler = (activity: IActivity) => {
-        const isActivityMessage: boolean = activity?.type === Constants.message;
-        const isHistoryMessage: boolean = isActivityMessage && (activity?.channelData?.tags?.includes(Constants.historyMessageTag) || activity?.channelData?.fromList);
+        raiseMessageEvent(activity,);
```
Comment (xTEddie) [inline chat-widget/src/plugins/newMessageEventHandler.ts:24]: nit: Remove extra `,` in the function parameter

File: chat-widget/src/plugins/newMessageEventHandler.ts [hunk 8]
```diff
@@ -1,112 +1,140 @@
...
+
+        const newMessageReceivedEvent: ICustomEvent = {
+            eventName: BroadcastEvent.HistoryMessageReceived,
+            payload: payload
+        };
+        BroadcastService.postMessage(newMessageReceivedEvent);
+
+        if (!isHistoryMessageReceivedEventRasied) {
+            // this is needed for reload scenarios, it helps to identify the last message received before the reload
+            isHistoryMessageReceivedEventRasied = true;
```
Comment (xTEddie) [inline chat-widget/src/plugins/newMessageEventHandler.ts:69]: typo: `isHistoryMessageReceivedEventRasied` -> `isHistoryMessageReceivedEventRaised`

File: chat-widget/src/firstresponselatency/Constants.ts (hunk 1 above)
Comment (elopezanaya) [reply chat-widget/src/firstresponselatency/Constants.ts:7]: MessagePayload can be extended, I'd like to keep the name since it shares type with ACS incoming messages and FRL

File: chat-widget/src/firstresponselatency/Constants.ts (hunk 2 above)
Comment (elopezanaya) [reply chat-widget/src/firstresponselatency/Constants.ts:25]: Its the only tracker, I dont see the need to be specific

File: chat-widget/src/firstresponselatency/FirstResponseLatencyTracker.ts (hunk 3 above)
Comment (elopezanaya) [reply chat-widget/src/firstresponselatency/FirstResponseLatencyTracker.ts:41]: This is one of those cases, where separation offers better readability

File: chat-widget/src/plugins/newMessageEventHandler.ts (hunk 8 above)
Comment (elopezanaya) [reply chat-widget/src/plugins/newMessageEventHandler.ts:69]: in my defense, it was cut&paste :P

**Review by xTEddie — APPROVED** [review]

**Review by Sonam-Ligade — APPROVED** [review]

**Review by mithunjohny — APPROVED** [review]
